README.md
## 👾 ![Logo] 👾

*A Display Add-on for the **[MiSTer FPGA]**.*

<br>

<img src = Pictures/tty2oled_video.gif align = left>

<br>
<br>
<br>

<div align = right>
  <div align = center>
  
This project provides both the **Hardware** as well as the <br>
**Software** to drive an **OLED** display with your Arduino <br>
with simple **USB** connection to your **MiSTer** system.
  
  </div>
</div>

<br>
<br>
<br>

---

<div align = center>

  **⸢ [Documentation] ⸥**
  **⸢ [Creating Graphics] ⸥**
  
</div>

---
[![Badge License]][License]
<!----------------------------------------------------------------------------->

[Badge License]: https://img.shields.io/badge/License-GPLv3-blue.svg?style=for-the-badge

[Logo]: Pictures/tty2oled_logo_120x46_blue_black.png?raw=true

[MiSTer FPGA]: https://github.com/MiSTer-devel

[Creating Graphics]: https://github.com/venice1200/MiSTer_tty2oled/wiki/Build_GSC_Gimp
[Documentation]: https://github.com/venice1200/MiSTer_tty2oled/wiki
[License]: LICENSE


MiSTer_SSD1322_USB/changelog.md
*************************** tty2oled Arduino-FW Changelog ***********************

  2023-07-02
  -Release to Stable

..2023-04-27
  -Cosmetics

  2023-04-20/2023-03-16
  -Add Command CMDSTTYACK to (set) enable(1)/disable(0) sending "ttyack". Enabled by default.
   Using no Command Delay (cDelay) if "ttyack" is not sent.

  2023-01-08
  -Add Boolean "runsTesting"

  2022-12-09..2023-01-07
  -Add more "hasXX"

  2022-12-08
  -Change 8266 NodeMCU Tilt Pin 16 (D0) to Pin 10 (SD3) to prevent false Rotation detection

  2022-11-27
  -Switch dtiv Version handling from "EEPROM" to "Preferences"

  2022-11-23
  -Removed unused Board Types

  2022-11-12..13
  -Add Suport for ESP32-S3 (ESP32-S3-DevKitC-1)
  -Removed OTA Support
  -Add EEPROM Support for d.ti Boards (IDE: ESP32* Dev Module)

  2022-10-30
  ***** Release to stable 221030 ******

  2022-10-24
  -Change Picture Array logoBin from "malloc" to "normal" defined type.

  2022-10-20
  -Add randomly chosen animated ScreenSaver after Corechange

  2022-10-15
  -Add short delay after Rotation-Pin Initialization (try to prevent false Rotation detection on 8266)
  -Add ScreenSaver Flying Toasters by Phillip Burgess/Adafruit
   https://learn.adafruit.com/animated-flying-toaster-oled-jewelry/code (MIT License) 
   Screensaver Bits: Bit 0=tty2oled Logo (Value 1), 1=MiSTer Logo (2), 2=Core Logo (4), 3=Time (8), 4=Date (16), 5=Stars (32), 6=Toast (64)

  2022-10-14
  -Changed filename logo.h to bitmaps.h
  
  2022-10-12
  -Modifying "Stars", adding Star-Depth-Color calculation

  2022-10-09
  ***** Release to stable 221009 ******

  2202-10-07
  -Fix for 8266 which don't get the Starfield Screensaver to prevent stability issues

  2022-09-30
  -Add the StarField Simulation as ScreenSaver
   Code from https://github.com/sinoia/oled-starfield (MIT License) 
   8266 must run at 160MHz!
   Screensaver Bits: Bit 0=tty2oled Logo (Value 1), 1=MiSTer Logo (=Value 2), 2=Core Logo (Value 4), 3=Time (Value 8), 4=Date (Value 16), 5=Stars (Value32)

  2022-09-25..27
  -New Command "CMDSWSAVER,[0/1]" for just Switching the Screensaver on/off
  -Workaround for the detection of PCA9536 = d.ti Board v1.2
  
  2022-09-02
  -New functions "oled_showcenterredtext" and "oled_setfont"

  2022-09-01
  -More usage of "dtiv"

  2022-08-26..28
  -New Command "CMDSHSYSHW" shows Hardware & Software Infos on Screen

  2022-09-18
  -Shorten the Date Format for the Screensaver. e.g. "19. September 2022" => "19-Sep-22".

  2022-08-07..12
  -Add RGB LED to Startup Screen (d.ti Board 1.2)
  -New Variable "dtiv" showing d.ti Board Revisions: 11=1.1 12=1.2...

  2022-08-23
  -Replace some "if (pcaAvail)" with "if (dtiv==12)"

  2022-07-31
  -SSD1322 Minor change
  -New Option "XOTA" for disable/enable the OTA functionality, Sketch consumption 65%->28%

  2022-07-23 
  ***** Release ***** to stable 220723 (based on 220714T)

  2022-07-12..14
  -Introduce ScreenSaver "Screens" which can be selected separatly (bitwise)
   Bit 0=tty2oled Logo (Value 1), 1=MiSTer Logo (=Value 2), 2=Core Logo (Value 4), 3=Time (Value 8), 4=Date (Value 16)
   Examples: 12=4+8=Core+Time, 5=1+4=tty2oled+Core, 13=1+4+8=tty2oled+Core+Time

  2022-07-09
  -Minor Update to make the Sketch compatible to ESP32 Board-Pack v2.0.4

  2022-07-08 -OBSOLETE- see 2022-07-12
  -Change ScreensaverColor to ScreeensaverMode
   Modes: 1: Show tty2oled Logo
          2: Show tty2oled Logo, Mister Logo
          3: Show tty2oled Logo, Mister Logo, Core Logo
          4: Show tty2oled Logo, Mister Logo, Core Logo, Time
          5: Show tty2oled Logo, Mister Logo, Core Logo, Time & Date

  2022-07-06
  -Adding two new fonts u8g2_font_commodore64_tr and u8g2_font_8bitclassic_tf

  2022-06-08
  -Note: Use U8g2_for_Adafruit_GFX setForegroundColor and setBackgroundColor together for inverted text
   See: https://github.com/olikraus/U8g2_for_Adafruit_GFX#command-reference

  2022-06-02
  -Add Function Prototypes for C++ Compatibilty

  2022-05-21
  -Only Cosmetics

  2022-05-07
  ** Release to stable**
  -Fix CMDPNOTE/CMDPTONE
  -Add CMDSHTIME
  -A lot Cosmetics
  -Lib Update: ESP32 Package 2.0.1 -> 2.0.3
  -Lib Update: Adafruit GFX 1.10.14 -> 1.11.1

  2022-04-28
  -Remove CMDSTEMP command and temnperature loop
  -Add CMDSHTEMP Command, show only once the actual MIC184 Temperature, for DTI Boards only
  

  2022-04-22
  -Command CMDSETTIME can be called by all MCU's but only the ESP32 is setting his internal RTC

  2022-04-18
  -Add Time to ScreenSaver but only if Time was set before, and only for ESP32

  2022-04-17
  -Add Code for new Commadn CMDSETTIME (stolen from tty2tft)
  -Add CMDNULL (Test) Command, for Troubleshooting and ACK Testing.
  
  2022-04-16
  -Fix CMDPNOTE/CMDPTONE Bug if Parameter is missing

  2022-04-10..13
  -Add Command "CMDSECD,[value]" to set the Command Delay
  -Add Command "CMDSHCD" to show the actual Command Delay "On Screen" (for Debugging)
  -Set cDelay for ESPDEV=15, Lolin32=60, 8266=60 after some manual tests with MiSTer SAM

  2022-04-07
  -Testing without "serial.flush" after "ttyack;"
  
  2022-04-07
  -Renamed Command CMDPTONE to CMDPNOTE
   >> CMDPNOTE,C,4,150,30,C,4,160,30,C,4,150,30,F,4,150,30,C,4,150,30  
  -Renamed Command CMDPFREQ to CMDPTONE

  2022-04-05
  -Wemos Lolin 32 Profile: #define cDelay 100 fixes MiSTer SAM issues with waitforack

  2022-04-01
  -Consolidate CMDSSCP/CMDSSCP2 and oled_showSmallCorePicture/oled_showSmallCorePictureV2

  2022-03-31
  ** Release **

  2022-03-23
  -Change cDelay from Variable to MCU dependent #define Value
   Fixes a response issue using "waitforack" and MiSTer_SAM with NodeMCU 8266

  2022-03-19
  -Screensaver Mod, use avarage of 2x2 Pixels for the small GSC Picture

  2022-03-16
  -Screensaver Mod, use 1/4 size Core Picture for the ScreenSaver
  -New Command CMDSSCP

  2022-03-02
  -Use more PicType "NONE" (Cosmetics)

  2022-02-17
  -Change minimum SceeenSaver Intervall from 10s to 5s

  2022-02-09 
  ** Release to Stable **

  2022-02-07

tty2oled.sh
#!/bin/bash

# By venice & ojaksch
#
# 2021-01-12 First release
# 2021-01-15 Adding debug to /tmp/tty2oled, device check else part
# 2021-01-16 Change send command "echo -ne ${newcore} >" to "echo ${newcore} >" without -ne to send "\n" (newline)
#            Arduino uses now "Serial.readStringUntil('\n');"
#            Feels more responsive as Serial Read on Arduino is not waitung for the timeout (1000ms)
#            Add "raw" to tty Parameter (see stty manpage)
# 2021-01-17 Add check for readable file "/tmp/CORENAME"
# 2021-01-19 Add "First Transmission" to clear send buffer (Preventing  weird issues after PowerOn)
# 2021-02-07 Changed Speed from 9600 to 57600
# 2021-02-14 Change Timed Loop to " inotifywait -e modify "/tmp/CORENAME" ".
#            Makes it much more responsive :-)
# 2021-03-29 USB Transfer realized by ojaksch, many thanks.
#            Modified Arduino Sketch and "tty2oled" for the USB Version of tty2oled.
#            All XBM must be stored on the MiSTer in the folder "/media/fat/tty2oledpics".
#            Instead of sending the Corename the Content of the XBM is send if an XBM exists. Done with the function "senddata".
#            "senddata" checks for an existing Picture-Folder and if the Folder exists the Picture-Data (if found) or the Corename are sent.
#            If the Folder does not exist, it's assumed the "SD Version" is used and only the corename is sent.
#            Serial Interface set to 115200 Baud now.
# 2021-04-12 Added an INI file
# 2021-05-05 Added sending Contrast Value
# 2021-05-14 Adding Command Line Parameter for Testing
#            Parameter 1: /dev/ttyUSBx Device
#            Parameter 2: SD or USB Mode
#            Parameter 3: Baudrate
#            Example: /usr/bin/tty2oled /dev/ttyUSB1
#            Example: /usr/bin/tty2oled /dev/ttyUSB1 SD
#            Example: /usr/bin/tty2oled /dev/ttyUSB1 USB 921600
# 2021-06-22 New Command Mode (Testing)
#            "CMCOR,[Corename]", "CMDCON,[Contrast]",  "CMDTEX,[Parameter]", "CMDGEO,[Parmeter]", "CMDRST", "CMDOTA"
# 2021-06-24 Adding folder "/media/fat/Scripts/.tty2oled/pics_pri"
#            Pictures found in this folder will be used before the "common" pictures.
#            The User himself is responsible for the content in this Folder.
# 2021-06-30 Added new INI Option/Command for Display Rotation ("CMDROT,[Parameter]")
# 2021-07-10 Fix Baudrate CLI Parameter handling
#            Change CLI Parameter order
#            Parameter 1: /dev/ttyUSBx Device
#            Parameter 2: Baudrate
#            Parameter 3: SD or USB Mode
#            Example: /usr/bin/tty2oled /dev/ttyUSB1
#            Example: /usr/bin/tty2oled /dev/ttyUSB1 115200
#            Example: /usr/bin/tty2oled /dev/ttyUSB1 921600 USB
# 2021-07-14 Clean up Script
# 2021-09-09 Moved from /etc/init.d to /media/fat/tty2oled
# 2021-09    Grayscale pictures implemented
# 2021-10-02 Complete rework of senddata
# 2021-10-05 USE_RANDOM_ALT to choose between _altX pictures
# 2022-01-23 Bugfix: Comment the reload of INI Files in function "senddata" because command line parameter don't work
# 2022-02-10 Added ScreenSaver functionality
# 2022-04-09 Make the Daemon more quiet
# 2022-04-24 Add "settime"
# 2022-06-17 Redo/Rework of PID and inotify
# 2022-07-22 New Screensaver Mode Handling
# 2022-09-20 Support for MiSTer SAM adding the tty2oled "SleepMode"
#            Create the file "/tmp/tty2oled_sleep" by using "touch /tmp/tty2oled_sleep" and the tty2oled Dameon goes to sleep.
#            Remove the file and the tty2oled Daemon goes back to work.
#
#

. /media/fat/tty2oled/tty2oled-system.ini
. /media/fat/tty2oled/tty2oled-user.ini
cd /tmp


# Debug function
dbug() {
  if [ "${debug}" = "true" ]; then
    if [ ! -e ${debugfile} ]; then # log file not (!) exists (-e) create it
      echo "---------- tty2oled Debuglog ----------" >${debugfile}
    fi
    echo "${1}" >>${debugfile} # output debug text
    echo "${1}"
  fi
}

# Send Contrast-Data function
sendcontrast() {
  if [ "${USBMODE}" = "yes" ]; then # Check the tty2xxx mode
    dbug "Sending: CMDCON,${CONTRAST}"
    echo "CMDCON,${CONTRAST}" >${TTYDEV} # Send Contrast Command and Value
    sleep ${WAITSECS}
  else
    echo "att" >${TTYDEV}       # Send an "att" to the MiSTer annoucing another Command
    sleep ${WAITSECS}           # sleep needed here ?!
    echo "CONTRAST" >${TTYDEV}  # Send "CONTRAST" annoucing the OLED Contrast Data as next
    sleep ${WAITSECS}           # sleep needed here ?!
    echo ${CONTRAST} >${TTYDEV} # Send Contrast Value
  fi
}

# Send Screensaver function
sendscreensaver() {
  if [ "${SCREENSAVER}" = "yes" ]; then # Check screensaver mode

    SCREENSAVER_MODE="0"
    [ "${SCREENSAVER_SCREEN_TTY2OLED}" = "yes" ] && let SCREENSAVER_MODE=${SCREENSAVER_MODE}+1
    [ "${SCREENSAVER_SCREEN_MISTER}" = "yes" ] && let SCREENSAVER_MODE=${SCREENSAVER_MODE}+2
    [ "${SCREENSAVER_SCREEN_CORE}" = "yes" ] && let SCREENSAVER_MODE=${SCREENSAVER_MODE}+4
    [ "${SCREENSAVER_SCREEN_TIME}" = "yes" ] && let SCREENSAVER_MODE=${SCREENSAVER_MODE}+8
    [ "${SCREENSAVER_SCREEN_DATE}" = "yes" ] && let SCREENSAVER_MODE=${SCREENSAVER_MODE}+16
    # echo "ScreenSaverMode: ${SCREENSAVER_MODE}"
    dbug "Sending: CMDSAVER,${SCREENSAVER_MODE},${SCREENSAVER_IVAL},${SCREENSAVER_START}"
    echo "CMDSAVER,${SCREENSAVER_MODE},${SCREENSAVER_IVAL},${SCREENSAVER_START}" >${TTYDEV} # Send Screensaver Command and Values

  else
    dbug "Sending: CMDSAVER,0,0,0"
    echo "CMDSAVER,0,0,0" >${TTYDEV} # Send Screensaver Command and Values
  fi
  sleep ${WAITSECS}
}

# Rotate Display function
sendrotation() {
  if [ "${USBMODE}" = "yes" ]; then # Check the tty2xxx mode
    if [ "${ROTATE}" = "yes" ]; then
      dbug "Sending: CMDROT,1"
      echo "CMDROT,1" >${TTYDEV} # Send Rotation if set to "yes"
      sleep ${WAITSECS}
      echo "CMDSORG" >${TTYDEV} # Show Start Screen rotated
      sleep 4
    #else
    #  dbug "Sending: CMDROT,0" > ${TTYDEV}
    #  echo "CMDROT,0" > ${TTYDEV}						# No Rotation
    #  sleep ${WAITSECS}
    #  echo "CMDSORG" > ${TTYDEV}						# Show Start Screen rotated
    #  sleep 4
    fi
  fi
}

# USB Send-Picture-Data function
senddata() {
  newcore="${1}"
  unset picfnam
  if [ "${USBMODE}" = "yes" ]; then                       # Check the tty2xxx mode
    if [ -e "${picturefolder_pri}/${newcore}.gsc" ]; then # Check for _pri pictures
      picfnam="${picturefolder_pri}/${newcore}.gsc"
    elif [ -e "${picturefolder_pri}/${newcore}.xbm" ]; then
      picfnam="${picturefolder_pri}/${newcore}.xbm"
    else
      picfolders="gsc_us xbm_us gsc xbm xbm_text" # If no _pri picture found, try all the others
      [ "${USE_US_PICTURE}" = "no" ] && picfolders="${picfolders//gsc_us xbm_us/}"
      [ "${USE_GSC_PICTURE}" = "no" ] && picfolders="${picfolders//gsc_us/}" && picfolders="${picfolders//gsc/}"
      [ "${USE_TEXT_PICTURE}" = "no" ] && picfolders="${picfolders//xbm_text/}"
      for picfolder in ${picfolders}; do
        for ((c = "${#newcore}"; c >= 1; c--)); do                                   # Manipulate string...
          picfnam="${picturefolder}/${picfolder^^}/${newcore:0:$c}.${picfolder:0:3}" # ...until it matches something
          [ -e "${picfnam}" ] && break
        done
        [ -e "${picfnam}" ] && break
      done
    fi
    if [ -e "${picfnam}" ]; then               # Exist?
      if [ "${USE_RANDOM_ALT}" = "yes" ]; then # Use _altX pictures?
        SAVEIFS="${IFS}"
        IFS=$'\n'
        ALTPICNUM=$(find $(dirname "${picfnam}") -name $(basename "${picfnam%.*}_alt")* | wc -l)
        IFS="${SAVEIFS}"
        if [ "${ALTPICNUM}" -gt "0" ]; then             # If more than 0 _altX pictures
          ALTPICRND=$((${RANDOM} % $((ALTPICNUM + 1)))) # then dice between 0 and count of found _altX pictures
           [ "${ALTPICRND}" -gt 0 ] && picfnam="${picfnam%.*}_alt"${ALTPICRND}".${picfolder:0:3}"
        fi # If 0 then original picture, otherwise _altX
      fi
      dbug "Sending: CMDCOR,${1},${TRANSITION}"
      echo "CMDCOR,${1},${TRANSITION}" >${TTYDEV}    # Send CORECHANGE" Command and Corename
      sleep ${WAITSECS}                              # sleep needed here ?!
      tail -n +4 "${picfnam}" | xxd -r -p >${TTYDEV} # The Magic, send the Picture-Data up from Line 4 and proces
    else                                             # No Picture available!
      echo "${1}" >${TTYDEV}                         # Send just the CORENAME
    fi                                               # End if Picture check
  else                                               # SD/Standard Mode ? Just send the Corename
    echo "${1}" >${TTYDEV}                           # Instruct the device to load the appropriate picture from SD card
  fi
}

sendtime() {
  timeoffset=$(date +%:::z)
  localtime=$(date '-d now '${timeoffset}' hour' +%s)
  echo "CMDSETTIME,${localtime}" >${TTYDEV}
  sleep ${WAITSECS}
}

# ** Main **
# Check for Command Line Parameter
if [ "${#}" -ge 1 ]; then # Command Line Parameter given, override Parameter
  #echo -e "\nUsing Command Line Parameter"
  dbug "\nUsing Command Line Parameter"
  TTYDEV=${1}                                             # Set TTYDEV with Parameter 1
  if [ -n "${2}" ]; then                                  # Parameter 2 Baudrate
    BAUDRATE=${2}                                         # Set Baudrate
  fi                                                      # end if Parameter 3
  if [ -n "${3}" ]; then                                  # Parameter 3 SD or USB Mode
    if [ "${3}" = "SD" ]; then                            # Parameter 3 = "SD" ?
      USBMODE="no"                                        # Un-Set USBMODE
    elif [ "${3}" = "USB" ]; then                         # Parameter 3 = "USB" ?
      USBMODE="yes"                                       # Set USBMODE
    else                                                  # Parameter not "USB" or "SD"!
      echo "Parameter 3 invalid"                          # Parameter 3 wrong
    fi                                                    # end if USB Mode
  fi                                                      # end if Parameter 3
  echo "Using Interface: ${TTYDEV} with ${BAUDRATE} Baud" # Device Output
  echo "USBMODE: ${USBMODE}"                              # Mode Output
fi                                                        # end if command line Parameter

# Let's go
if [ -c "${TTYDEV}" ]; then # check for tty device
  dbug "${TTYDEV} detected, setting Parameter: ${BAUDRATE} ${TTYPARAM}."
  stty -F ${TTYDEV} ${BAUDRATE} ${TTYPARAM} # set tty parameter
  sleep ${WAITSECS}
  echo "QWERTZ" >${TTYDEV} # First Transmission to clear serial send buffer
  dbug "Send QWERTZ as first transmission"
  sleep ${WAITSECS}
  sendcontrast												# Set Contrast
  sendrotation												# Set Display Rotation
  sendtime													# Set time and date
  sendscreensaver											# Set Screensaver
  while true; do											# main loop
    if [ -r ${corenamefile} ]; then							# proceed if file exists and is readable (-r)
      if [ -f ${SLEEPFILE} ]; then							# Sleepmode = Yes
        dbug "The tty2oled daemon is sleeping!"
        #inotifywait -qq -t 60 -e delete "${SLEEPFILE}"		  # Sleepmode is waiting it here for 60 secs
        inotifywait -qq -e delete "${SLEEPFILE}"		  # Sleepmode is waiting it here
      fi
      if [ ! -f ${SLEEPFILE} ]; then				  # Sleepmode = No
        newcore=$(<${corenamefile})				  # get CORENAME
        #if [ "$newcore" != "$oldcore" ]; then
          dbug "Read CORENAME: -${newcore}-"
          dbug "Send -${newcore}- to ${TTYDEV}."
          senddata "${newcore}" 				   # The "Magic"
          oldcore=$newcore
          if [ "${debug}" = "false" ]; then
            inotifywait -qq -e modify "${corenamefile}"            # wait here for next change of corename, -qq for quietness
          elif [ "${debug}" = "true" ]; then
            inotifywait -e modify "${corenamefile}"                # but not -qq when debugging
          fi
	#else
        #  dbug "Core not changed!"
        #fi #newcore != oldcore
      fi
    else # CORENAME file not found
      dbug "File ${corenamefile} not found!"
    fi # end if /tmp/CORENAME check
  done # end while
else   # no tty detected
  echo "No ${TTYDEV} Device detected, abort."
  dbug "No ${TTYDEV} Device detected, abort."
fi # end if tty check
# ** End Main **
